data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/codec/parquet/S3OutputStream.java

```diff
@@ -147,36 +166,48 @@ public void write(byte[] byteArray, int o, int l) {
     public void flush() {
     }
 
-    @Override
-    public void close() {
+    public CompletableFuture<?> close(final Consumer<Boolean> runOnCompletion, final Consumer<Throwable> runOnError) {
         if (open) {
             open = false;
-            possiblyStartMultipartUpload();
-            if (position > 0) {
-                uploadPart();
-            }
+            try {
+                possiblyStartMultipartUpload();
+
+                if (position > 0) {
+                    uploadPart();
+                }
+
+                CompletedPart[] completedParts = new CompletedPart[etags.size()];
+                for (int i = 0; i < etags.size(); i++) {
+                    completedParts[i] = CompletedPart.builder()
+                            .eTag(etags.get(i))
+                            .partNumber(i + 1)
+                            .build();
+                }
 
-            CompletedPart[] completedParts = new CompletedPart[etags.size()];
-            for (int i = 0; i < etags.size(); i++) {
-                completedParts[i] = CompletedPart.builder()
-                        .eTag(etags.get(i))
-                        .partNumber(i + 1)
+                LOG.debug("Completing S3 multipart upload with {} parts.", completedParts.length);
+
+                CompletedMultipartUpload completedMultipartUpload = CompletedMultipartUpload.builder()
+                        .parts(completedParts)
                         .build();
-            }
+                CompleteMultipartUploadRequest completeMultipartUploadRequest = CompleteMultipartUploadRequest.builder()
+                        .bucket(bucket)
+                        .key(key)
+                        .uploadId(uploadId)
+                        .multipartUpload(completedMultipartUpload)
+                        .build();
+                CompletableFuture<CompleteMultipartUploadResponse> multipartUploadResponseCompletableFuture = s3Client.completeMultipartUpload(completeMultipartUploadRequest);
 
-            LOG.debug("Completing S3 multipart upload with {} parts.", completedParts.length);
-
-            CompletedMultipartUpload completedMultipartUpload = CompletedMultipartUpload.builder()
-                    .parts(completedParts)
-                    .build();
-            CompleteMultipartUploadRequest completeMultipartUploadRequest = CompleteMultipartUploadRequest.builder()
-                    .bucket(bucket)
-                    .key(key)
-                    .uploadId(uploadId)
-                    .multipartUpload(completedMultipartUpload)
-                    .build();
-            s3Client.completeMultipartUpload(completeMultipartUploadRequest);
+                multipartUploadResponseCompletableFuture.join();
+
+                runOnCompletion.accept(true);
+                return multipartUploadResponseCompletableFuture;
```

Comment:
The execution is complete in line 200. Why do we want to return the future here ?

Reply:
We can return null here